### ISSUE TITLE
Migrate to UIkit 3

### ISSUE THREAD

**Issue by eSilverStrike** [Geeklog-Core/geeklog#748]
It is in beta at the time of writing this issue (UIKit website already shows UIKit 3 first) but by the time Geeklog 2.2.0 is released it will be out so we might as well migrate over.

There is a migration script tool to help find the changes that need to happen in the markup.

This migration could break some plugins template files. I wonder if we should add some sort of version information for plugins to check (could happen in the plugins functions.php file for it's templates) what version of UIkit is being used. Not much we could do at this point except maybe add a note to the error log. Depending on what has changed between major versions of UIKit everything could work fine (or not).

**Comment by eSilverStrike** [comment]
@mystralkk just wondering on how we should handle this. I was initially thinking we just update the denim theme. The problem with this is that plugins also now have template specific themes so there is the potential of a plugin like the forum which currently doesn't support uikit3 loading in the original uikit library files instead. Uikit2 and Uikit3 do not play nicely together. Plus there are classes and attributes are different and need to be changed in the individual template files depending on the uikit version.

This will also create issues with child themes (like denim curve) which is based on the original denim theme.

I see 2 options. 

**The first option** is to put some sort of library version requirements (for uikit) in the theme functions.php file. Changes would have to be done to the Geeklog scripts class as well (and maybe elsewhere) so plugins could check for the right uikit version and if not found, fall back to the default theme template files. Template files could also possible support 2 versions of uikit if we add the appropriate if statements where needed.

**The second option** is easier. We create a brand new theme and call it denim3. This theme is just all the orginal denim theme files but with the required changes to make it compatible with uikit3. This way plugins would have to specifically support the theme else it would fall back to the default template files. Child themes would continue to work as well.

**Comment by mystralkk** [comment]
I would choose option 2, too.  Option 1 may be cleaner, but it really is more confusing for users in that they have to specify the UIkit version number in functions.php.

**Comment by eSilverStrike** [comment]
Okay I've scratched the surface of migrating the denim theme. I have added the denim3 theme which is just a straight copy of the denim theme (along with a few initial changes). I've started the migration process by using the browser bookmark option found here: https://getuikit.com/docs/migration

LOTS still needs to be updated including the functions.php. I was hoping @dengen would be available to help out as css is not my forte. I've decided to commit the initial changes as I am hoping a few of us can work on this. It's not just swapping in and out different css classes in the templates as I was hoping.

@mystralkk How did we decide to deal with the css source files again? Some of the css that deal with ids and classes needs to be modified and or removed. We might as well start doing this the right way...

I figure we update the theme first before moving to adding the denim3 templates to the core plugins.

It also looks like we may need to update core a bit as well. For example the Switch Language Block. For forms like this we just needed to add the uk-form class to the form tag for UiKit2 which we did in the template. Now that class is removed in UiKit3  and now we need to add a uk-select class to the select tag which is automatically built by Geeklog.

**Comment by mystralkk** [comment]
Since @dengenxp is unlikely to help us any time soon (we still haven't got a reply from him), I think we have to do without him.  About CSS source files, I haven't changed anything with the Scripts class.  CSS files specified in functions.php and by calling $_SCRIPTS->setCSSFile or $_SCRIPTS->setCSS methods will be concatenated and minified automatically.

**Comment by eSilverStrike** [comment]
Sorry @mystralkk I meant the css souce:  https://github.com/Geeklog-Core/geeklog/tree/master/public_html/layout/denim3/css_src

Looks like @dengenxp  was using the following to compile the css files

https://nodejs.org/en/
http://stylus-lang.com/

I've never used the above before (I've always just edited css files directly). I know we talked about this before but I forgot what we had decided on. Are these the tools you want to use (or Sass, less)? Might as well start things off the right way :-)

My current plan is just to continue to update the template files replacing the uikit classes as I go and keeping temporarily most of dengens custom classes for uikit2. Once that is done then we can go in start tweaking/adding/removing our custom css classes as needed.

**Comment by mystralkk** [comment]
I used node.js a bit but haven't used Stylus at all.  What I suggested last time was to use tools provided at http://beautifytools.com/ and convert all Stylus files to CSS files and then the resulting CSS files to SASS files.  If I were to migrate to UIkit 3, I would use Sass (or LESS) files, since UIkit use Sass and LESS.

**Comment by eSilverStrike** [comment]
@mystralkk @dengenxp It looks like UIkit 3 is going to force us to remove the rest of the embedded HTML in Geeklog. This is mainly input controls (like check boxes, radio options, select dropdowns) and a few buttons (mostly from admin lists). The reason for this is Uikit 3 now wants us to add classes to these controls to control their styling (before we just added a class to the form tag). 

I am thinking the best way to do this is to add a new template file called commoncontrols.thtml and within the file we have blocks for each control that can be reused. So the commoncontrols.html would look like this (I haven't added all the controls yet):

```
{# begin {templatelocation} - Geeklog Common Controls #}

<!-- BEGIN type-hidden -->
<input type="hidden" name="{name}" value="{value}"{XHTML}>
<!-- END type-hidden -->

<!-- BEGIN type-checkbox -->
<input type="checkbox" name="{name}" title="{title}"{!if style} style="{style}"{!endif} class="uk-checkbox {class}"{!if onchange} onchange="{onchange}"{!endif}{!if onclick} onclick="{onclick}"{!endif}{XHTML}>
<!-- END type-checkbox -->

<!-- BEGIN type-select -->
<select name="{name}"{!if style} style="{style}"{!endif} class="uk-select uk-form-width-small {class}"{!if onchange} onchange="{onchange}"{!endif}{!if onclick} onclick="{onclick}"{!endif}>{select_items}</select>
<!-- END type-select -->

```

What do you think? Is there a better option?

**Comment by eSilverStrike** [comment]
And then in the code (for example admin/article.php) you would replaced

```
    $filter = $LANG_ADMIN['topic']
        . ': <select name="tid" style="width: 125px" onchange="this.form.submit()">'
        . $seltopics . '</select>';
```

with 

```
    $tcc = COM_newTemplate($_CONF['path_layout']);
    $tcc->set_file(array('commoncontrols' => 'commoncontrols.thtml'));
    $tcc->set_block('commoncontrols', 'type-select'); 
    
    $tcc->set_var('name', 'tid');
    $tcc->set_var('onchange', 'this.form.submit()');
    $tcc->set_var('select_items', $seltopics);
    
    $filter = $LANG_ADMIN['topic'] . ": " . $tcc->parse('commoncontrols', 'type-select', true);
```

**Comment by dengenxp** [comment]
I do not know yet if your proposal is the best solution.
A few days ago, in order to grasp the point at issue, I started a work to apply UIKit3 to Denim.

**Comment by eSilverStrike** [comment]
OK. Do you mean by keeping Denim for Uikit 2 and by having Denim3 support Uikit 3? The reason I did that was non core plugins like for example the Forum which support the Denim theme assume it uses Uikit 2. I also did it this way for child themes. Child themes assume their parent theme will always support the version of uikit it was created with.

Did you want to continue to work on Denim3? I am more than happy to let you do that if you wish?

I can still continue on and move the rest of the hardcoded html code out of Geeklog and into template files (this should have been done a while ago). The only html I will probably leave in will be the hidden inputs since it will never need to be styled.

**Comment by dengenxp** [comment]
I agree with the idea of having Denim and Denim 3 coexist.
The current progress is about 60 to 70%. I'm thinking I will continue until it is completed.
I would be happy if you continue to move the hard-coded HTML code into template files.

**Comment by eSilverStrike** [comment]
Sounds good

**Comment by eSilverStrike** [comment]
@dengenxp can you take a look at changeset 02f1a7b761f172cdfbdfd731fe892b32c1cf89a9 and tell me what you think? I removed the hardcoded tables for the groups rights and made it a flex list.

I know flex lists are not supported by older browsers but they just fall back to a regular list.

I've included the style below in the actual template files for now.
```
<style>
.flex-rights-list {
  display: flex;
  flex-wrap: wrap;
  list-style-type: none;
}

.flex-rights-list > li {
  width: 300px;
}
</style>
```
If it is a style we want to keep how do you want me to add it to the Denim theme? Just use the custom.css file and then you will move it over?

**Comment by dengenxp** [comment]
I think that it was improved. I have already merged this into the Denim_Three theme.
The development of the Denim_Three theme has almost completed.
In order to receive the evaluation of you and others, I will commit the Denim_Tree theme to the master repository.

**Comment by eSilverStrike** [comment]
Sounds good. What do you think about the Modern Curve theme and the css. I need to update it with the above but we have  the etag issue from #725. I think I will remove etag support from Modern Curve  and just merge all the Modern Curve theme css files into 2, 1 for ltr and 1 for rtl. Are you fine with that as well?

**Comment by eSilverStrike** [comment]
I'll also remove the denim3 theme then from the repository.

**Comment by eSilverStrike** [comment]
Hey @dengenxp, I notice that their are no denim_three templates for the core plugins. This means that for the denim_three theme the core plugins end up using the default templates which are meant more for themes like Modern Curve which are not using a ui library like Uikit. 

To fix this in most cases we could probably just create a new directory called denim_three in the plugins templates directory and copy over the contents of the denim directory. The main things we would have to change are the stuff dealing with the inputs and forms. I already got started and added some stuff for the Polls plugin to fix the results page, etc...

**Comment by eSilverStrike** [comment]
I've updated the templates for demin_three  for the polls and staticpage. For the editors I am not sure if you eventually want to follow the layouts as designed here https://getuikit.com/docs/form#layout but I kept it similar to how you did the article editor templates. I hope that is okay.

**Comment by dengenxp** [comment]
I appreciate your good work.

The current style using dl, dd, dt tags was originally from the professional_css theme. To be honest, I did not like it much, but I had to adapt to it. Perhaps people will say that they should use the style of UIkit this time! I think a little that it is selfish. Even now, it has become a responsive style sufficiently.

But, I am also thinking that I can change it. I'd like to change it altogether when we separate the admin page.

**Comment by eSilverStrike** [comment]
Sounds good regarding the current styling for the editor and submit forms. I've updated the Calendar but it's styling for editor and submit uses tables and divs as well. I've made a few small changes but I am just going to leave that for now and get back to removing the hardcoded html.

**Comment by eSilverStrike** [comment]
@dengen @mystralkk Just an update on what I am doing for removing the hardcoded HTML. I am only doing it for the Geeklog Core code and Core plugins. Obviously I am not touching any vender code (third party classes, etc..). I also don't plan on updating the Geeklog installer at this time. After any changes I make I do test everything to make sure it works as it should.

These are the items I usually am coming across and fixing as I go:

```
<input type="checkbox"
<select>
<input type="image" 
<input type="submit"
<input type="text"
<input type="file"
<form>
```
I am currently leaving all hidden inputs since these cannot be styled:

`<input type="hidden"`

What do you think I should do with things like:

```
<div dir="ltr" class="allowed_autotags">
<span class="warningsmall">
```
and variations of the above with div and span tags? Should I use the common/controls.thtml template file and create separate blocks for each or just leave them as is?

For functions like COM_whatsNewBlock and similar hardcoded HTML I will use a template files.

**Comment by dengenxp** [comment]
I will respect your policy. If problems arise, let's fix it at that point.

**Comment by eSilverStrike** [comment]
Hey @dengenxp 

I think we have removed most of the hardcoded HTML.

Most of the inputs have been converted (except hidden) and most of the regular HTML has been removed. I left the div and span tags that get set custom classes but removed others that use common classes and added them to common.thtml or created there own template files. I left a bit in the trackback code but I have a feeling we will be removing trackback and pingbacks in the next version of Geeklog (see #836).

The only thing I didn't touch was the config.class.php for the Geeklog Configuration. It has a fair bit of HTML in it and 2 different versions based on $_CONF['supported_version_theme']. Should we only make this support the latest version or leave as is?

The other thing I left was the installer. At some point we should update it to use Uikit 3 but that can be for another version of Geeklog (#849)
